typing/rec_check.ml

```diff
@@ -577,15 +577,30 @@ let rec expression : Typedtree.expression -> term_judg =
       *)
       expression arg << Guard
     | Texp_apply (e, args)  ->
+        (* [args] may contain omitted arguments, corresponding to labels in
+           the function's type that were not passed in the actual application.
+           The arguments before the first omitted argument are passed to the
+           function immediately, so they are dereferenced. The arguments after
+           the first omitted one are stored in a closure, so guarded.
+           The function itself is called immediately (dereferenced) if there
+           is at least one argument before the first omitted one, stored in
+           the closure (guarded) otherwise. *)
+        let rec split_args before_omitted = function
+          | (_, None) :: rest -> before_omitted, rest
+          | ((_, Some _) as arg) :: rest ->
+            split_args (arg :: before_omitted) rest
+          | [] -> before_omitted, []
+        in
```

Comment:
I am not fond of the way this function is written:

- it is not obvious what `before_omitted` means (it could be "the arguments before the first omitted arguments", which you mean, but it could also be "the omitted arguments that we have seen before the current point"); I propose to fix this by having a non-tail-recursive function without this accumulator parameter
- the `dereferenced` arguments are returned in reverse order; this is innocuous to compute the whole result but it still looks like a potential bug on first reading, I prefer to preserve the order or point out explicitly. I propose to fix this by having a non-tail-recursive function.
- the first omitted argument is silently dropped, when the function name suggest that it keeps all arguments. This does not make a difference in the result but then again it increases the potential for confusion. Please keep it as part of the second list returned.

```ocaml
let rec split_args = function
| [] -> [], []
| ((_, None) :: _) as omitted -> [], omitted
| ((_, Some _) as arg) :: rest ->
    let applied, omitted = split_args rest in
    arg :: applied, omitted
in
let applied, omitted = split_args args in
...
```

Reply:
To me, "omitted" means the argument corresponds to a label that was not given. So I want a distinction between the omitted argument, which has no corresponding expression, and the delayed arguments, which can have expressions that will be guarded. But I agree that it feels weird that some omitted args end up in the delayed part but not the first one.
How about that:
```ocaml
let rec split_args ~has_omitted_arg = function
| [] -> [], []
| (_, None) :: rest -> split_args ~has_omitted_arg:true rest
| (_, Some arg) :: rest ->
  let applied, delayed = split_args ~has_omitted_arg rest in
  if has_omitted_arg then applied, arg :: delayed
  else arg :: applied, delayed
in
let applied, delayed = split_args ~has_omitted_arg:false args in
...
```
I could also tweak the name `split_args` to make it clear it drops the ones without an actual expression (`split_actual_args`, for instance).

Reply:
Yes, I like your suggestion as well. Having the `delayed` args be expression instead of expression options will also simplify the rest of the code.

Nitpick: I would break a line after `if has_omitted_arg` because the `then` and `else` cases are symmetric so it is nice to keep them aligned.